base_layer/p2p/src/services/liveness/mod.rs

```diff
@@ -20,28 +20,131 @@
 //  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 //  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
+//! # Liveness Service
+//!
+//! This service is responsible for sending pings to any peer as well as maintaining
+//! some very basic counters for the number of ping/pongs sent and received.
+//!
+//! It consists of:
+//! - A service handle which makes requests to the Liveness backend. Types of requests can be found in the
+//!   [LivenessRequest] enum.
+//! - A handler for incoming [PingPong] messages.
+//!
+//! In future, this service may be expanded to included periodic pings to maintain
+//! latency and availability statistics for peers.
+//!
+//! [LivenessRequest]: ./messages/enum.LivenessRequets.html
+//! [PingPong]: ./messages/enum.PingPong.html
+
 mod error;
+mod handler;
 mod messages;
 mod service;
 mod state;
 
-use self::messages::{LivenessRequest, LivenessResponse};
+use self::{error::LivenessError, handler::LivenessHandler, service::LivenessService, state::LivenessState};
 use crate::{
-    executor::transport::{self, Requester, Responder},
+    executor::{
+        transport::{self, Requester},
+        ServiceInitializationError,
+        ServiceInitializer,
+    },
     services::{
-        liveness::{error::LivenessError, service::LivenessService, state::LivenessState},
-        ServiceHandles,
+        comms_outbound::CommsOutboundHandle,
+        domain_deserializer::DomainMessageDeserializer,
+        ServiceHandlesFuture,
+        ServiceName,
     },
+    tari_message::{NetMessage, TariMessageType},
 };
-use std::sync::Arc;
+use futures::{future, Future, Stream};
+use log::*;
+use std::{fmt::Debug, sync::Arc};
+use tari_comms::inbound_message_service::InboundTopicSubscriber;
+
+pub use self::messages::{LivenessRequest, LivenessResponse, PingPong};
+use tari_comms::domain_subscriber::MessageInfo;
+
+pub type LivenessHandle = Requester<LivenessRequest, Result<LivenessResponse, LivenessError>>;
+
+const LOG_TARGET: &'static str = "base_layer::p2p::services::liveness";
+
+/// Initializer for the Liveness service handle and service future.
+pub struct LivenessInitializer {
+    inbound_subscriber: Arc<InboundTopicSubscriber<TariMessageType>>,
```

Comment:
Ye this is service dependent - so you can do it in any way you like - however I'd like to keep service specific code in the initializer for that service rather than passing in `LivenessInitializer::new(comms.inbound_subcriber().subscription(TariMessageType::new(NetMessage::PingPong))`

You can do `LivenessInitializer::new(comms.inbound_subscriber())` - might even just pass in comms and leave the linking of comms to the service completely up to the service initializer implementation.

Reply:
I think pass in the Comms

Reply:
Done - although the downside is that this requires a full comms stack to test. I've added a function to create the LivenessInitializer from a subscriber. In the near future, we probably want to make it easy to setup a full (development) comms stack in integration tests

Reply:
Making a mock Comms stack that just has the Subscriber and Future based interface should be pretty easy.